Commander: start pulling arming related parts into separate folder

MaEtUgR opened the pull request:
**Describe problem solved by this pull request**
Goal: The user knows in advance if he'll be able to arm.

**Describe your solution**
First step: Pull things that relate to arming logic into the folder Arming and make sure to have a clear interface for all checks that happen before arming can take place

**Describe possible alternatives**
Hack my way through commander trying to run important bits and pieces in a random and unmaintainable way.

**Test data / coverage**
Only draft yet but I'm constantly testing every single step by using the existing unit-sitl tests and my own sitl testing. Hopefully I try to get to the point where efficient gtest unit testing is possible.

**Additional Context**
#7055

jlecoeur commented:
> Goal: The user knows in advance if he'll be able to arm.

That is a revolution

ndepal commented:
Related? https://github.com/PX4/Firmware/pull/13025 https://github.com/PX4/Firmware/pull/12989

MaEtUgR commented:
@ndepal Definitely related. Thanks for referencing. I can rebase either way no matter which one goes first.

dagar commented:
Glad to see some movement here!

One thing I'd like to get sorted out is the line between preflight, prearm, and checks that can run continuously.

I'm wondering if we can get to the point where nearly everything results in status flags and have the arming machine act on that data rather than actually executing the checks. 

The other missing piece is fitting this into mavlink.

dagar commented on src/modules/commander/Arming/PreFlightCheck/PreFlightCheck.cpp:
```diff
@@ -691,286 +1226,37 @@ static bool failureDetectorCheck(orb_advert_t *mavlink_log_pub, const vehicle_st
-	}
-
-	/* ---- BARO ---- */
-	if (checkSensors) {
```
Perhaps not what you want to tackle here, but I'm wondering if all this sensors code should be owned by the sensors module. Then a higher level sensors status would be published for commander to act on.

MaEtUgR commented:
@dagar Thanks for looking over it and your input! I try to share as many thoughts as possible. I want to first get it all in a less interdependent and hence more maintainable adaptable state.

> I'd like to get sorted out is the line between preflight, prearm

My intentions after talking to a bunch of people about this: Preflight checks are the ones that determine if you can fly or not (user-focused), the pre arm checks (I left the name for now but it might not make sense anymore) are just one kind of preflight check since they prevent arming and hence flying.

> and checks that can run continuously.
> I'm wondering if we can get to the point where nearly everything results in status flags and have the arming machine act on that data rather than actually executing the checks.

Exactly the intention and why I started refactoring this. The goal is that there is no check that was not done before giving an arm command/input, no surprises to the user anymore. The arming state machine itself just knows if there's green light.

> The other missing piece is fitting this into mavlink.

I agree, first I have to leave the strings but have them all optional (to not spam), then I plan to fill the flags at the beginning of https://mavlink.io/en/messages/common.html#SYS_STATUS as in-between solution and eventually we'll have the events interface taking care of delivering the status to the UI.

> Perhaps not what you want to tackle here, but I'm wondering if all this sensors code should be owned by the sensors module. Then a higher level sensors status would be published for commander to act on.

I thought exactly the same let's go in steps.

My currently intended steps:
- pull all checks together into PreFlightCheck into as small of an include and symbol scope as possible, dependencies get clear
- make sure checks are callable without output and with one single call covering all desired checks
- refactor code of checks, I saw they add up to >1200 lines of code which is almost not possible without duplication and there are better places for certain checks e.g. where the data of interest is produced or consumed

This PR should only do the first point otherwise it gets too big of a change.

MaEtUgR commented:
@dagar @bresch I rebased on master with the only relatively trivial conflict of these lines:
https://github.com/PX4/Firmware/commit/549fb0d5de8a8eaebaa54d6e4d458876b53b60d2#diff-a2bcb238e25d8a97ab2aee48f54dbc48R497-R513

It would be nice to get this in soon. I can rebase other conflicting prs easily.

dagar commented on src/modules/commander/Arming/PreFlightCheck/PreFlightCheck.hpp:
```diff
@@ -0,0 +1,64 @@
+				const safety_s &safety, const uint8_t arm_requirements);
+
+	typedef enum {
+		ARM_REQ_NONE = 0,
```
Note for later - PreFlightCheck could own these parameters (COM_ARM_MIS_REQ, COM_ARM_WO_GPS, etc).

MaEtUgR replied:
Definitely and wrap them without external bit level access being necessary.

dagar commented on src/modules/commander/Arming/PreFlightCheck/PreFlightCheck.cpp:
```diff
@@ -45,15 +45,11 @@
-		if (reportFailures) {
-			mavlink_log_critical(mavlink_log_pub, "Arming denied! Flying with USB is not safe");
-		}
+		mavlink_log_critical(mavlink_log_pub, "Arming denied! Flying with USB is not safe");
```
I might be missing some of the high level design, but isn't this potentially a step in the wrong direction?

Don't we ultimately want to be able to know pre-arm check status continuously, and separately get the relevant messages as needed?

MaEtUgR replied:
Yes, it's just the start that looks into the wrong direction becuase the reportFailures flag was a local variable that was never changed and I plan to only have the output once you know what checks have failed and not spread over all the checks.

I can revert this change since it's anyways not how it ends up to be and limit the focus of this pr on the raw structure.

MaEtUgR replied:
I removed the functional changes since this should be pure refactoring and the functional changes should be a real step in the right direction before merging.

dagar commented on src/modules/commander/Arming/PreFlightCheck/checks/CMakeLists.txt:
```diff
@@ -0,0 +1,45 @@
+#
+############################################################################
+
+target_sources(PreFlightCheck PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/preArmCheck.cpp)
```
This is kind of weird. Why not just add them higher level or make a single library and link it?

MaEtUgR replied:
It is a single Library and gets linked. The files are just located in a subfolder. Should I reference them all in the main CMakeLists of the library? That would probably help easy understanding compared to adding sources in the subfolder.

MaEtUgR replied:
I added them to the main library definition in the last commit.

dannyfpv commented:
tested on pixhawk4 v5 f-450
                 pixhawk v4pro f-450
<preflight fail: heading estimate not stable>
change to master firmware  not issues

<img width="761" alt="Screen Shot 2019-10-29 at 10 59 57 AM" src="https://user-images.githubusercontent.com/31113129/67812134-97f69b80-fa5b-11e9-81c0-73614af8ea0f.png">

MaEtUgR commented:
> Preflight fail: heading estimate not stable, unable to arm.

This was a bug introduced with https://github.com/PX4/Firmware/pull/13036 and fixed in https://github.com/PX4/Firmware/pull/13261 and https://github.com/PX4/Firmware/pull/13267. I was unlucky to have master at the exact time it was broken as my base. I'll rebase and this will fix the problem.

MaEtUgR commented:
Thanks for the reviews and comments also from @bkueng.
I rebased, removed the funtional commit since like @dagar mentioned they don't really help by themselfes yet and addressed all the comments about includes with respect to folders in the last commit.

MaEtUgR commented:
I'm assuming CI problems are now unrelated. It's ready for review check.

dagar commented:
Looks good to me. Can you rebase?

MaEtUgR commented:
Rebased
The conflict was just these four space characters: https://github.com/PX4/Firmware/commit/a475d71ca975c3343f56eb91412231486e014604#diff-eb94b84d1c6e5b1530f8ad834cef9b22R3761

MaEtUgR commented:
Fixed style using rebase in https://github.com/PX4/Firmware/compare/856bca21538ee02c66fde813d666d3eebac6e75d..fbaee7624805c46309318d20fa2815ad34dcdc1b

MaEtUgR commented:
Upon request of @dagar I did a quick test on a real vehicle with a pixhawk 4 mini. I can normally arm and fly when all checks pass. When the vehicle is upside down I can't arm because of the failing roll check.
